hail/src/main/scala/is/hail/expr/types/physical/PBetterLocus.scala

```diff
@@ -0,0 +1,131 @@
+package is.hail.expr.types.physical
+
+import is.hail.annotations._
+import is.hail.asm4s._
+import is.hail.expr.ir.{EmitClassBuilder, EmitCodeBuilder, EmitMethodBuilder}
+import is.hail.expr.types.virtual.{TArray, TString}
+import is.hail.utils.FastIndexedSeq
+import is.hail.variant._
+
+object PBetterLocus {
+  def apply(rg: ReferenceGenome): PLocus = PBetterLocus(rg.broadcastRG)
+
+  def apply(rg: ReferenceGenome, required: Boolean): PLocus = PBetterLocus(rg.broadcastRG, required)
+}
+
+final case class PBetterLocus(rgBc: BroadcastRG, required: Boolean = false) extends PLocus {
+  def rg: ReferenceGenome = rgBc.value
+
+  def _asIdent = "locusplus"
+
+  // 32 high bits are the contig, 32 low bits is the position
+  val representation: PInt64 = PInt64(required)
+
+  override def _pretty(sb: StringBuilder, indent: Call, compact: Boolean): Unit = sb.append(s"PBetterLocus($rg)")
+
+  private def contigIndex(value: Long): Int = (value >>> 32).toInt
+
+  def contig(value: Long): String = rg.contigs(contigIndex(value))
+
+  def contigType: PString = PCanonicalString(required = true)
+
+  def positionType: PInt32 = PInt32(true)
+
+  def codeOrdering(mb: EmitMethodBuilder[_], other: PType): CodeOrdering = {
+    assert(other isOfType this)
+    new CodeOrderingCompareConsistentWithOthers {
+      type T = Long
+
+      override def compareNonnull(lhs: Code[Long], rhsc: Code[Long]): Code[Int] = {
+        val rhs = other match {
+          case _: PBetterLocus => rhsc
+          case t: PCanonicalLocus => t.toBetterLocus(mb, rhsc)
+        }
+
+        Code.invokeStatic2[java.lang.Long, Long, Long, Int]("compare", lhs, rhs)
+      }
+    }
+  }
+
+  def setRequired(required: Boolean): PType = if (this.required == required)
+    this
+  else
+    PBetterLocus(this.rgBc, required)
+
+  private[physical] def codeContigs(ecb: EmitClassBuilder[_]): PIndexableValue = {
+    ecb.addLiteral(
+      rg.contigs,
+      PType.literalPType(TArray(TString), rg.contigs.toSeq)
```

Comment:
Use `toFastIndexedSeq`.  Two reasons: one, the literal needs to be an IndexedSeq, not a Seq.  Two, toSeq in some cases used to recons an Array into a List!  So bad.  I checked the latest version of Scala and it does the right thing now, but toSeq leaves a very bad taste in my mouth.